File: 2018/day01-chronal-calibration/calibrate.py
#!/usr/bin/env python

from typing import List


def main() -> None:
    with open("day1.in") as f:
        freqChanges = list(map(int, f.readlines()))

    print(f"Part 1: {getSumFrequencies(freqChanges)}")
    print(f"Part 2: {getFirstDuplicate(freqChanges)}")


def getSumFrequencies(freqChanges: List[int]) -> int:
    return sum(freqChanges)


def getFirstDuplicate(freqChanges: List[int]) -> int:
    total, seen, i = 0, set(), 0
    while True:
        total += freqChanges[i]
        if total in seen:
            return total
        seen.add(total)
        i = (i + 1) % len(freqChanges)


if __name__ == '__main__':
    main()

File: 2018/day01-chronal-calibration/calibrate.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	freqChanges := parseInput("day1.in")
	fmt.Printf("Part 1: %d\n", getSumFrequencies(freqChanges))
	fmt.Printf("Part 2: %d\n", getFirstDuplicate(freqChanges))

}

func parseInput(filename string) []string {
	bytes, _ := ioutil.ReadFile(filename)
	lines := strings.Split(string(bytes), "\n")
	return lines[:len(lines)-1]
}

func getSumFrequencies(freqChanges []string) int {
	sum := 0
	for _, change := range freqChanges {
		val, _ := strconv.Atoi(change)
		sum += val
	}
	return sum
}

func getFirstDuplicate(freqChanges []string) int {
	sum, seen := 0, make(map[int]bool)
	for i := 0; ; i = (i + 1) % len(freqChanges) {
		val, _ := strconv.Atoi(freqChanges[i])
		sum += val

		if _, duplicate := seen[sum]; duplicate {
			return sum
		}
		seen[sum] = true
	}
}
